logging/net_appender.go

```diff
@@ -206,40 +207,45 @@ func (nl *NetAppender) backgroundWorker() {
 	}
 }
 
-// Sync will flush the internal buffer of logs.
-func (nl *NetAppender) Sync() error {
-	for {
-		batch := func() []*apppb.LogEntry {
-			nl.toLogMutex.Lock()
-			defer nl.toLogMutex.Unlock()
-
-			if len(nl.toLog) == 0 {
-				return nil
-			}
-
-			batchSize := writeBatchSize
-			if len(nl.toLog) < writeBatchSize {
-				batchSize = len(nl.toLog)
-			}
+// Returns whether there is more work to do or if an error was encountered while trying to ship logs
+// over the network.
+func (nl *NetAppender) syncOnce() (bool, error) {
+	nl.toLogMutex.Lock()
+	defer nl.toLogMutex.Unlock()
 
-			ret := nl.toLog[:batchSize]
-			nl.toLog = nl.toLog[batchSize:]
+	if len(nl.toLog) == 0 {
+		return false, nil
+	}
 
-			return ret
-		}()
+	batchSize := writeBatchSize
+	if len(nl.toLog) < writeBatchSize {
+		batchSize = len(nl.toLog)
+	}
 
-		if len(batch) == 0 {
-			return nil
-		}
+	batch := nl.toLog[:batchSize]
+	err := nl.remoteWriter.write(batch)
+	if err == nil {
+		// On success remove the batch from the queue.
+		nl.toLog = nl.toLog[batchSize:]
+		return len(nl.toLog) > 0, nil
+	} else {
+		// On error, abort the sync attempt. But keep the logs to sync in the queue. Such that a
+		// follow-up sync call does not miss logs.
+		return false, err
+	}
+}
 
-		err := nl.remoteWriter.write(batch)
+// Sync will flush the internal buffer of logs.
+func (nl *NetAppender) Sync() error {
+	for {
+		moreToDo, err := nl.syncOnce()
 		if err != nil {
-			// On an error, the failed batch gets put on the back of the queue. Logs can be sent out
-			// of order. We depend on log front-ends to sort the results by time if they care about
-			// order.
-			nl.addBatchToQueue(batch)
```

Comment:
Post-patch -- `addBatchToQueue` is no longer used. I was debating removing it. It's a little tricky to rewrite and I wasn't convinced its utility couldn't be used as a public facing function.

Reply:
> I was debating removing it.

If it's now not used apart from tests should I just make a quick followup PR to remove it? Or was there another reason you wanted to keep it around?